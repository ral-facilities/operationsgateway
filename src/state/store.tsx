import {
  configureStore,
  combineReducers,
  PreloadedState,
} from '@reduxjs/toolkit';
import OperationsGatewayMiddleware, {
  listenToMessages,
} from './middleware/operationsgateway.middleware';
import configReducer from './slices/configSlice';
import tableReducer from './slices/tableSlice';
import searchReducer from './slices/searchSlice';
import plotReducer from './slices/plotSlice';

const rootReducer = combineReducers({
  config: configReducer,
  table: tableReducer,
  search: searchReducer,
  plots: plotReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(OperationsGatewayMiddleware),
  });
}

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];

export const store = setupStore();

listenToMessages(store.dispatch);
