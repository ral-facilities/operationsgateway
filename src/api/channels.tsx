import React from 'react';
import axios, { AxiosError } from 'axios';
import {
  Channel,
  FullChannelMetadata,
  FullScalarChannelMetadata,
  Record,
} from '../app.types';
import { useQuery, UseQueryResult, UseQueryOptions } from 'react-query';
import { Column } from 'react-table';
import { roundNumber } from '../table/cellRenderers/cellContentRenderers';
import { selectUrls } from '../state/slices/configSlice';
import { useAppSelector } from '../state/hooks';

export const generateChannelMetadata = (
  records: Record[]
): FullChannelMetadata[] => {
  const metadata: FullChannelMetadata[] = [];

  records.forEach((record: Record) => {
    const keys = Object.keys(record.channels);
    keys.forEach((key: string) => {
      if (!metadata.find((channel) => channel.systemName === key)) {
        const channel: Channel = record.channels[key];
        const channelDataType = channel.metadata.channel_dtype;
        const newMetadata: FullChannelMetadata = {
          systemName: key,
          channel_dtype: channelDataType,
        };
        metadata.push(newMetadata);
      }
    });
  });

  return metadata;
};

// TODO change this when we have a proper channel info endpoint to query
// This just fetches metadata from the records endpoint at the moment
const fetchChannels = (apiUrl: string): Promise<FullChannelMetadata[]> => {
  return axios.get(`${apiUrl}/records`).then((response) => {
    const records: Record[] = response.data;
    const metadata = generateChannelMetadata(records);
    return metadata;
  });
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export const useChannels = <T extends unknown = FullChannelMetadata[]>(
  options?: UseQueryOptions<FullChannelMetadata[], AxiosError, T, string[]>
): UseQueryResult<T, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery(
    ['channels'],
    (params) => {
      return fetchChannels(apiUrl);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      ...(options ?? {}),
    }
  );
};

export const constructColumns = (channels: FullChannelMetadata[]): Column[] => {
  const myColumns: Column[] = [
    {
      accessor: 'timestamp',
      Header: 'Timestamp',
    },
    {
      accessor: 'shotnum',
      Header: 'Shot Number',
    },
    {
      accessor: 'activeArea',
      Header: 'Active Area',
    },
    {
      accessor: 'activeExperiment',
      Header: 'Active Experiment',
    },
  ];

  channels.forEach((channel: FullChannelMetadata) => {
    const newColumn: Column = {
      Header: () => {
        const headerName = channel.userFriendlyName
          ? channel.userFriendlyName
          : channel.systemName;
        // Provide an actual header here when we have it
        // TODO: do we need to split on things other than underscore?
        const parts = headerName.split('_');
        const wordWrap = parts.map(
          (part, i) =>
            // \u200B renders a zero-width space character
            // which allows line-break but isn't visible
            part + (i < parts.length - 1 ? '_\u200B' : '')
        );
        return <React.Fragment>{wordWrap.join('')}</React.Fragment>;
      },
      accessor: channel.systemName,
      // TODO: get these from data channel info
      channelInfo: channel,
    };
    if (channel.channel_dtype === 'scalar') {
      newColumn.Cell = ({ value }) =>
        typeof value === 'number' &&
        typeof channel.significantFigures === 'number' ? (
          <React.Fragment>
            {roundNumber(
              value,
              channel.significantFigures,
              channel.scientificNotation ?? false
            )}
          </React.Fragment>
        ) : (
          <React.Fragment>{String(value ?? '')}</React.Fragment>
        );
    }
    myColumns.push(newColumn);
  });
  return myColumns;
};

export const getScalarChannels = (
  channels: FullChannelMetadata[]
): FullScalarChannelMetadata[] => {
  return channels.filter(
    (channel) => channel.channel_dtype === 'scalar'
  ) as FullScalarChannelMetadata[];
};

const useAvailableColumnsOptions = {
  select: (data: FullChannelMetadata[]) => constructColumns(data),
};

const useScalarChannelsOptions = {
  select: (data: FullChannelMetadata[]) => getScalarChannels(data),
};

export const useAvailableColumns = (): UseQueryResult<Column[], AxiosError> => {
  return useChannels(useAvailableColumnsOptions);
};

export const useScalarChannels = (): UseQueryResult<
  FullScalarChannelMetadata[],
  AxiosError
> => {
  return useChannels(useScalarChannelsOptions);
};
