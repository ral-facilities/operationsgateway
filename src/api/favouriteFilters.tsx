import {
  useMutation,
  UseMutationResult,
  useQuery,
  useQueryClient,
  UseQueryResult,
} from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { FavouriteFilter, FavouriteFilterPost } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

const addFavouriteFilter = (
  apiUrl: string,
  favouriteFilter: FavouriteFilterPost
): Promise<string> => {
  const queryParams = new URLSearchParams();

  queryParams.append('name', favouriteFilter.name);
  queryParams.append('filter', favouriteFilter.filter);

  return axios
    .post<string>(
      `${apiUrl}/users/filters`,
      {},
      {
        params: queryParams,
        headers: {
          Authorization: `Bearer ${readSciGatewayToken()}`,
        },
      }
    )
    .then((response) => response.data);
};

export const useAddFavouriteFilter = (): UseMutationResult<
  string,
  AxiosError,
  FavouriteFilterPost
> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (favouriteFilter: FavouriteFilterPost) =>
      addFavouriteFilter(apiUrl, favouriteFilter),
    onError: (error) => {
      console.log('Got error ' + error.message);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['favouriteFilters'] });
    },
  });
};

const fetchFavouriteFilters = (apiUrl: string): Promise<FavouriteFilter[]> => {
  return axios
    .get(`${apiUrl}/users/filters`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useFavouriteFilters = (): UseQueryResult<
  FavouriteFilter[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['favouriteFilters'],

    queryFn: () => {
      return fetchFavouriteFilters(apiUrl);
    },
  });
};
