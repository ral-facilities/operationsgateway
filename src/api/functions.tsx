import axios, { AxiosError } from 'axios';
import { DataType, FunctionItem, FunctionToken } from '../app.types';
import { useQuery, UseQueryResult } from '@tanstack/react-query';
import { selectUrls } from '../state/slices/configSlice';
import { useAppSelector } from '../state/hooks';
import { readSciGatewayToken } from '../parseTokens';
import { Token } from '../filtering/filterParser';

const fetchTokens = (apiUrl: string): Promise<Token[]> => {
  return axios
    .get(`${apiUrl}/functions/tokens`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data.map((token: FunctionToken) => ({
        label: token.name,
        value: token.symbol,
        type: 'string', // TODO sort this
      }));
    });
};

const fetchFunctionReturnType = (
  apiUrl: string,
  index: number,
  functionItems: FunctionItem[]
): Promise<DataType> => {
  const queryParams = new URLSearchParams();
  const pairs: string[][] = [];
  functionItems.forEach((functionItem, i) => {
    if (i === index) {
      let expression = '';
      functionItem.expression.forEach((token) => {
        expression += token.value;
      });
      queryParams.append(
        'function',
        JSON.stringify({ ...functionItem, expression: expression })
      );
    }
    pairs.push([functionItem.name, functionItem.returnType ?? '']);
  });

  queryParams.append(
    'function_types',
    JSON.stringify(Object.fromEntries(pairs))
  );
  return axios
    .get(`${apiUrl}/functions`, {
      params: queryParams,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data as DataType;
    });
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export const useTokens = <T extends unknown = Token[]>(): UseQueryResult<
  T,
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);
  return useQuery(
    ['tokens'],
    (params) => {
      return fetchTokens(apiUrl);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
    }
  );
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export const useFunctionReturnType = <T extends unknown = DataType>(
  index: number,
  functionItems: FunctionItem[]
): UseQueryResult<T, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);
  return useQuery(
    ['functionReturnType'],
    (params) => {
      return fetchFunctionReturnType(apiUrl, index, functionItems);
    // },
    // {
    //   onError: (error) => {
    //     console.log('Got error ' + error.message);
    //   },
    }
  );
};
