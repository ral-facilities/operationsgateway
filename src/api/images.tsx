import {
  UseQueryResult,
  keepPreviousData,
  useQuery,
} from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { APIFunctionState } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';
import { selectQueryParams } from '../state/slices/searchSlice';

export interface FalseColourParams {
  colourMap?: string;
  lowerLevel?: number;
  upperLevel?: number;
}

export interface ColourMapsParams {
  [category: string]: string[] | undefined;
}

export const fetchImage = async (
  apiUrl: string,
  recordId: string,
  channelName: string,
  functionsState: APIFunctionState,
  falseColourParams?: FalseColourParams
): Promise<string> => {
  const params = new URLSearchParams();

  if (falseColourParams) {
    const { colourMap, lowerLevel, upperLevel } = falseColourParams;

    if (typeof colourMap !== 'undefined')
      params.set('colourmap_name', colourMap);
    if (typeof lowerLevel !== 'undefined')
      params.set('lower_level', lowerLevel.toString());
    if (typeof upperLevel !== 'undefined')
      params.set('upper_level', upperLevel.toString());
  }

  if (!falseColourParams || params.toString().length === 0)
    params.set('original_image', 'true');

  functionsState.functions.forEach((func) => {
    params.append('functions', JSON.stringify(func));
  });

  return axios
    .get(`${apiUrl}/images/${recordId}/${channelName}`, {
      params,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
      responseType: 'blob',
    })
    .then((response) => {
      return URL.createObjectURL(response.data);
    });
};

export const fetchColourBar = async (
  apiUrl: string,
  falseColourParams: FalseColourParams
): Promise<string> => {
  const params = new URLSearchParams();
  const { colourMap, lowerLevel, upperLevel } = falseColourParams;

  if (colourMap) params.set('colourmap_name', colourMap);
  if (lowerLevel) params.set('lower_level', lowerLevel.toString());
  if (upperLevel) params.set('upper_level', upperLevel.toString());

  return axios
    .get(`${apiUrl}/images/colour_bar`, {
      params,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
      responseType: 'blob',
    })
    .then((response) => {
      return URL.createObjectURL(response.data);
    });
};

export const fetchColourMaps = async (
  apiUrl: string
): Promise<ColourMapsParams> => {
  return axios
    .get(`${apiUrl}/images/colourmap_names`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

interface CrosshairDimensionType {
  position: number;
  intensity: { x: number[]; y: number[] };
  fwhm: number;
}

interface CrosshairResponse {
  row: CrosshairDimensionType;
  column: CrosshairDimensionType;
}

// TODO: name
export const fetchCrosshair = async (
  apiUrl: string,
  recordId: string,
  channelName: string,
  functionsState: APIFunctionState,
  position?: { x: number; y: number }
): Promise<CrosshairResponse> => {
  const params = new URLSearchParams();

  functionsState.functions.forEach((func) => {
    params.append('functions', JSON.stringify(func));
  });
  if (position) params.set('position', `[${position.x},${position.y}]`);

  return axios
    .get(`${apiUrl}/images/${recordId}/${channelName}/crosshair`, {
      params,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useImage = (
  recordId: string,
  channelName: string,
  falseColourParams?: FalseColourParams
): UseQueryResult<string, AxiosError> => {
  const { functions } = useAppSelector(selectQueryParams);
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['images', recordId, channelName, functions, falseColourParams],

    queryFn: () => {
      return fetchImage(
        apiUrl,
        recordId,
        channelName,
        functions,
        falseColourParams
      );
    },

    // set to display old image whilst new one is loading
    placeholderData: keepPreviousData,
  });
};

export const useColourBar = (
  falseColourParams: FalseColourParams
): UseQueryResult<string, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['colourbar', falseColourParams],

    queryFn: () => {
      return fetchColourBar(apiUrl, falseColourParams);
    },

    // set to display old colour bar whilst new one is loading
    placeholderData: keepPreviousData,
  });
};

export const useColourMaps = (): UseQueryResult<
  ColourMapsParams,
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['colourmaps'],

    queryFn: () => {
      return fetchColourMaps(apiUrl);
    },
  });
};

export const useImageCrosshair = (
  recordId: string,
  channelName: string,
  position: { x: number; y: number } | undefined,
  enabled: boolean
): UseQueryResult<CrosshairResponse, AxiosError> => {
  const { functions } = useAppSelector(selectQueryParams);
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['imageCrosshair', recordId, channelName, position, functions],

    queryFn: () => {
      return fetchCrosshair(apiUrl, recordId, channelName, functions, position);
    },
    enabled,
  });
};
