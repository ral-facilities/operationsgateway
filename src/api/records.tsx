import axios, { AxiosError } from 'axios';
import { useQuery, UseQueryResult } from 'react-query';
import {
  Channel,
  DateRange,
  ImageChannel,
  Record,
  RecordRow,
  ScalarChannel,
  SortType,
  WaveformChannel,
} from '../app.types';
import { useAppSelector } from '../state/hooks';
import { selectQueryParams } from '../state/slices/searchSlice';
import { parseISO, format } from 'date-fns';
import { selectUrls } from '../state/slices/configSlice';

const fetchRecords = async (
  apiUrl: string,
  sort: SortType,
  dateRange: DateRange,
  offsetParams?: {
    startIndex: number;
    stopIndex: number;
  }
): Promise<Record[]> => {
  const params = new URLSearchParams();

  for (const [key, value] of Object.entries(sort)) {
    // API recognises sort values as metadata.key or channel.key
    // Therefore, we must construct the appropriate parameter
    const sortKey = [
      'timestamp',
      'shotnum',
      'activeArea',
      'activeExperiment',
    ].includes(key)
      ? `metadata.${key}`
      : `channels.${key}`;
    params.append('order', `${sortKey} ${value}`);
  }

  if (Object.keys(dateRange).length > 0) {
    const timestampObj = {
      $and: [
        {
          'metadata.timestamp': {
            $gt: dateRange.fromDate,
            $lt: dateRange.toDate,
          },
        },
      ],
    };
    params.append('conditions', JSON.stringify(timestampObj));
  }

  if (offsetParams) {
    params.append('skip', JSON.stringify(offsetParams.startIndex));
    params.append(
      'limit',
      JSON.stringify(offsetParams.stopIndex - offsetParams.startIndex + 1)
    );
  }

  return axios.get(`${apiUrl}/records`, { params }).then((response) => {
    const records: Record[] = response.data;
    return records;
  });
};

const fetchRecordCountQuery = (
  apiUrl: string,
  dateRange: DateRange
): Promise<number> => {
  const params = new URLSearchParams();
  if (Object.keys(dateRange).length > 0) {
    const timestampObj = {
      $and: [
        {
          'metadata.timestamp': {
            $gt: dateRange.fromDate,
            $lt: dateRange.toDate,
          },
        },
      ],
    };
    params.append('conditions', JSON.stringify(timestampObj));
  }

  return axios
    .get(`${apiUrl}/records/count`, { params })
    .then((response) => response.data);
};

export const useRecordsPaginated = (): UseQueryResult<
  RecordRow[],
  AxiosError
> => {
  const { page, resultsPerPage, sort, dateRange } =
    useAppSelector(selectQueryParams);
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery<
    Record[],
    AxiosError,
    RecordRow[],
    [
      string,
      {
        page: number;
        resultsPerPage: number;
        sort: SortType;
        dateRange: DateRange;
      }
    ]
  >(
    ['records', { page, resultsPerPage, sort, dateRange }],
    (params) => {
      const { page, sort, dateRange } = params.queryKey[1];
      // React Table pagination is zero-based
      const startIndex = page * resultsPerPage;
      const stopIndex = startIndex + resultsPerPage - 1;
      return fetchRecords(apiUrl, sort, dateRange, { startIndex, stopIndex });
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      select: (data) =>
        data.map((record: Record) => {
          const timestampString = record.metadata.timestamp;
          const timestampDate = parseISO(timestampString);
          const formattedDate = format(timestampDate, 'yyyy-MM-dd HH:mm:ss');
          const recordRow: RecordRow = {
            timestamp: formattedDate,
            shotnum: record.metadata.shotnum,
            activeArea: record.metadata.activeArea,
            activeExperiment: record.metadata.activeExperiment,
          };

          const keys = Object.keys(record.channels);
          keys.forEach((key: string) => {
            const channel: Channel = record.channels[key];
            let channelData;
            const channelDataType = channel.metadata.channel_dtype;

            switch (channelDataType) {
              case 'scalar':
                channelData = (channel as ScalarChannel).data;
                break;
              case 'image':
                channelData = (channel as ImageChannel).thumbnail;
                channelData = (
                  <img
                    src={`data:image/jpeg;base64,${channelData}`}
                    alt={key}
                    style={{ border: '1px solid #000000' }}
                  />
                );
                break;
              case 'waveform':
                channelData = (channel as WaveformChannel).thumbnail;
                channelData = (
                  <img
                    src={`data:image/jpeg;base64,${channelData}`}
                    alt={key}
                    style={{ border: '1px solid #000000' }}
                  />
                );
            }

            recordRow[key] = channelData;
          });

          return recordRow;
        }),
    }
  );
};

// TODO: should we integrate this with useRecordsPaginated? and have options to
// pass the recordRow select function and which of the query params to apply?
export const useRecords = (): UseQueryResult<Record[], AxiosError> => {
  const { page, resultsPerPage, sort, dateRange } =
    useAppSelector(selectQueryParams);
  const { apiUrl } = useAppSelector(selectUrls);
  return useQuery<
    Record[],
    AxiosError,
    Record[],
    [
      string,
      {
        page: number;
        resultsPerPage: number;
        sort: SortType;
        dateRange: DateRange;
      }
    ]
  >(
    ['records', { page, resultsPerPage, sort, dateRange }],
    (params) => {
      const { page, sort, dateRange } = params.queryKey[1];
      // React Table pagination is zero-based
      const startIndex = page * resultsPerPage;
      const stopIndex = startIndex + resultsPerPage - 1;
      return fetchRecords(apiUrl, sort, dateRange, { startIndex, stopIndex });
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
    }
  );
};

export const useRecordCount = (): UseQueryResult<number, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const { dateRange } = useAppSelector(selectQueryParams);

  return useQuery<
    number,
    AxiosError,
    number,
    [string, { dateRange: DateRange }]
  >(
    ['recordCount', { dateRange }],
    (params) => {
      const { dateRange } = params.queryKey[1];
      return fetchRecordCountQuery(apiUrl, dateRange);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
    }
  );
};
