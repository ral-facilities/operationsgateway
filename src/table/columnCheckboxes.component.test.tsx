import React from 'react';
import { Column } from 'react-table';
import ColumnCheckboxes from './columnCheckboxes.component';
import { screen, act } from '@testing-library/react';
import {
  flushPromises,
  getInitialState,
  renderComponentWithStore,
} from '../setupTests';
import { PreloadedState } from '@reduxjs/toolkit';
import { RootState } from '../state/store';
import { useAvailableColumns } from '../api/channels';

jest.mock('../api/channels', () => {
  const originalModule = jest.requireActual('../api/channels');

  return {
    __esModule: true,
    ...originalModule,
    useChannels: jest.fn(),
    useAvailableColumns: jest.fn(),
  };
});

describe('Column Checkboxes', () => {
  const availableColumns: Column[] = [
    {
      Header: 'Name',
      accessor: 'name',
    },
    {
      Header: 'Active Area',
      accessor: 'activeArea',
      channelInfo: {
        systemName: 'activeArea',
        channel_dtype: 'scalar',
      },
    },
    {
      Header: 'Active Experiment',
      accessor: 'activeExperiment',
      channelInfo: {
        systemName: 'activeExperiment',
        userFriendlyName: 'Active Experiment',
        channel_dtype: 'scalar',
      },
    },
  ];

  let state: PreloadedState<RootState>;

  const createView = (initialState = state) => {
    return renderComponentWithStore(<ColumnCheckboxes />, {
      preloadedState: initialState,
    });
  };

  beforeEach(() => {
    state = { ...getInitialState(), table: { ...getInitialState().table } };
    (useAvailableColumns as jest.Mock).mockReturnValue({
      data: availableColumns,
      isLoading: false,
    });
  });

  it('renders correctly when unchecked', () => {
    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders correctly when checked', () => {
    state.table.selectedColumnIds = availableColumns.map((col) => col.accessor);
    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('does not render a timestamp checkbox if a timestamp column exists', () => {
    const amendedColumns: Column[] = [
      ...availableColumns,
      {
        Header: 'Timestamp',
        accessor: 'timestamp',
      },
    ];
    (useAvailableColumns as jest.Mock).mockReturnValue({
      data: amendedColumns,
      isLoading: false,
    });

    createView();
    expect(screen.queryByLabelText('timestamp checkbox')).toBeNull();
  });

  it('sends selectColumn when checkbox is checked', async () => {
    const { store } = createView();
    await act(async () => {
      screen.getByLabelText('name checkbox').click();
      await flushPromises();
    });

    expect(store.getState().table.selectedColumnIds).toEqual(['name']);
  });

  it('calls onColumnClose when checkbox is unchecked', async () => {
    state.table.selectedColumnIds = availableColumns.map((col) => col.accessor);
    const { store } = createView();
    await act(async () => {
      screen.getByLabelText('name checkbox').click();
      await flushPromises();
    });
    expect(store.getState().table.selectedColumnIds).toEqual(
      availableColumns
        .filter((col) => col.accessor !== 'name')
        .map((col) => col.accessor)
    );
  });

  it('returns null if a column is not fully defined', () => {
    availableColumns[0].accessor = undefined;
    (useAvailableColumns as jest.Mock).mockReturnValue({
      data: availableColumns,
      isLoading: false,
    });

    createView();
    const checkboxes = screen.getAllByRole('checkbox');
    expect(checkboxes.length).toEqual(2);
    expect(screen.queryByText('name')).toBeNull();
  });

  it.todo('calls onChecked when checkbox is clicked via shift-click');
});
