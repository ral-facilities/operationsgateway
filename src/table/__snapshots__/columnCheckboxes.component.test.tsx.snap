// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column Checkboxes renders correctly when checked 1`] = `
<DocumentFragment>
  <div
    style="overflow: auto; height: 150px;"
  >
    <div
      style="height: 126px; width: 100%; position: relative;"
    >
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 42px; transform: translateY(0px);"
      >
        <label
          for="name"
        >
          name
        </label>
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            aria-label="name checkbox"
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            id="name"
            type="checkbox"
            value="name"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 42px; transform: translateY(42px);"
      >
        <label
          for="activeArea"
        >
          activeArea
        </label>
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            aria-label="activeArea checkbox"
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            id="activeArea"
            type="checkbox"
            value="activeArea"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 42px; transform: translateY(84px);"
      >
        <label
          for="activeExperiment"
        >
          Active Experiment
        </label>
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            aria-label="activeExperiment checkbox"
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            id="activeExperiment"
            type="checkbox"
            value="activeExperiment"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Column Checkboxes renders correctly when unchecked 1`] = `
<DocumentFragment>
  <div
    style="overflow: auto; height: 150px;"
  >
    <div
      style="height: 126px; width: 100%; position: relative;"
    >
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 42px; transform: translateY(0px);"
      >
        <label
          for="name"
        >
          name
        </label>
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            aria-label="name checkbox"
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            id="name"
            type="checkbox"
            value="name"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 42px; transform: translateY(42px);"
      >
        <label
          for="activeArea"
        >
          activeArea
        </label>
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            aria-label="activeArea checkbox"
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            id="activeArea"
            type="checkbox"
            value="activeArea"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 42px; transform: translateY(84px);"
      >
        <label
          for="activeExperiment"
        >
          Active Experiment
        </label>
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            aria-label="activeExperiment checkbox"
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            id="activeExperiment"
            type="checkbox"
            value="activeExperiment"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
